File: Data Structures & Algorithms/combination-target-sum-ii/submission-3.java
class Solution {
    List<List<Integer>> res;

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        res = new ArrayList<>();

        Arrays.sort(candidates);
        solve(candidates, 0, 0, target, new ArrayList<>());
        
        return res;
    }

    private void solve(int[] nums, int i, int sum, int target, List<Integer> list){
        if(target == sum){
            res.add(new ArrayList<>(list));
            return;
        }

        for(int j=i; j < nums.length; j++){
            if(j > i && nums[j] == nums[j-1]){
                continue;
            }

            if(sum + nums[j] <= target){
                list.add(nums[j]);
                solve(nums, j+1, sum + nums[j], target, list);
                list.remove(list.size() - 1);
            }
        }
    }
}

File: Data Structures & Algorithms/combination-target-sum/submission-3.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        res = []
        sub = []

        def solve(i, sum):
            if target == sum:
                res.append(sub.copy())
                return

            if i == len(nums) or sum > target:
                return

            if nums[i] + sum <= target:
                sub.append(nums[i])
                solve(i, sum + nums[i])
                sub.pop()
                solve(i+1, sum)
            
        solve(0, 0)
        return res



